llvm/lib/Analysis/IVDescriptors.cpp

```diff
@@ -741,10 +742,9 @@ RecurrenceDescriptor::isFindIVPattern(RecurKind Kind, Loop *TheLoop,
                                   : APInt::getMinValue(NumBits);
         ValidRange = ConstantRange::getNonEmpty(Sentinel + 1, Sentinel);
       } else {
-        assert(IsSigned && "Only FindFirstIV with SMax is supported currently");
-        ValidRange =
-            ConstantRange::getNonEmpty(APInt::getSignedMinValue(NumBits),
-                                       APInt::getSignedMaxValue(NumBits) - 1);
+        APInt Sentinel = IsSigned ? APInt::getSignedMaxValue(NumBits)
+                                  : APInt::getMaxValue(NumBits);
+        ValidRange = ConstantRange::getNonEmpty(Sentinel, Sentinel - 1);
```

Comment:
Is this right? I thought the point of the range was to exclude everything except the sentinel, but if I've understood ConstantRange::getNonEmpty correctly it creates the range [Lower, Upper) where in this case Lower == Sentinel. This means Sentinel is included in the range? This seems to contradict the findlastiv case where we create the range [Sentinel + 1, Sentinel)

Reply:
As a compromise, I've clarified it in the header comment and kept (Sentinel + 1, Sentinel) for uniformity.